caikit_nlp/toolkit/data_prep_utils.py

```diff
@@ -0,0 +1,163 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helper functions for preparing training data, e.g., preparing data formats.
+"""
+# Standard
+from copy import deepcopy
+from typing import Callable, Tuple
+
+# First Party
+from caikit.core.data_model import DataStream
+from caikit.core.toolkit import error_handler
+import alog
+
+# Local
+from ..data_model import GenerationTrainRecord
+from ..resources.pretrained_model import HFAutoCausalLM, HFAutoSeq2SeqLM
+from .verbalizer_utils import render_verbalizer
+
+log = alog.use_channel("DATA_PREP")
+error = error_handler.get(log)
+
+IGNORE_ID = -100
+
+
+def build_tokenize_function(
+    tokenizer: "AutoTokenizer",
+    max_source_length: int,
+    max_target_length: int,
+    verbalizer: str,
+    task_type: str,
+) -> Tuple[Callable, bool]:
+    """Builds tokenizer functions which can be mapped over train streams to process
+    data which can then be easily passed to a DataLoader.
+
+    Args:
+        tokenizer: AutoTokenizer
+            Model tokenizer to be used in preprocessing, i.e., when we iterate over our data.
+        max_source_length: int
+            Max length of sequences being considered.
+        max_target_length: int
+            Max length of target sequences being predicted.
+        verbalizer: str
+            Verbalizer template to be used for formatting data. This template may use brackets
+            to indicate where fields from the data model TrainGenerationRecord must be rendered.
+        task_type: str
+            Str indicating which task is being accomplished; currently used for determining
+            tokenization / preprocessing behavior.
+
+    Returns:
+        Tuple(Callable, bool)
+            Mappable tokenize function to be applied to a training stream and bool indicating
+            whether or not the stream needs to be unwrapped, i.e., each sample yields a stream
+            of 1+ samples.
+    """
+
+    def tokenize_function_language_model(
```

Comment:
Should we move the tokenization function to respective resources?

Reply:
ah yes, the static method route sound good!

Reply:
Cool sounds good - it's not beautiful, but I did refactor it + the tests into the resource classes 😄 may be a good idea to clean this up a bit over time, but at least we can preserve the resource APIs throughout future refactors on it